File: packages/cross_file/cross_file_android/android/src/main/kotlin/dev/flutter/packages/cross_file_android/ProxyApiRegistrar.kt
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package dev.flutter.packages.cross_file_android

import android.content.Context
import dev.flutter.packages.cross_file_android.proxies.AndroidLibraryPigeonProxyApiRegistrar
import dev.flutter.packages.cross_file_android.proxies.ContentResolverProxyApi
import dev.flutter.packages.cross_file_android.proxies.DocumentFileProxyApi
import dev.flutter.packages.cross_file_android.proxies.InputStreamProxyApi
import dev.flutter.packages.cross_file_android.proxies.PigeonApiContentResolver
import dev.flutter.packages.cross_file_android.proxies.PigeonApiDocumentFile
import dev.flutter.packages.cross_file_android.proxies.PigeonApiInputStream
import io.flutter.plugin.common.BinaryMessenger

/**
 * Implementation of [AndroidLibraryPigeonProxyApiRegistrar] that provides each ProxyApi
 * implementation and any additional resources needed by an implementation.
 */
open class ProxyApiRegistrar(binaryMessenger: BinaryMessenger, var context: Context) :
    AndroidLibraryPigeonProxyApiRegistrar(binaryMessenger) {

  override fun getPigeonApiDocumentFile(): PigeonApiDocumentFile {
    return DocumentFileProxyApi(this)
  }

  override fun getPigeonApiContentResolver(): PigeonApiContentResolver {
    return ContentResolverProxyApi(this)
  }

  override fun getPigeonApiInputStream(): PigeonApiInputStream {
    return InputStreamProxyApi(this)
  }
}

File: packages/cross_file/cross_file_darwin/darwin/Classes/ProxyAPIRegistrar.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Flutter
import Foundation

/// Implementation of `CrossFileDarwinApisPigeonProxyApiRegistrar` that provides any additional resources needed by API implementations.
open class ProxyAPIRegistrar: CrossFileDarwinApisPigeonProxyApiRegistrar {
  init(binaryMessenger: FlutterBinaryMessenger) {
    super.init(binaryMessenger: binaryMessenger, apiDelegate: ProxyAPIDelegate())
  }

  func createConstructorNullError(type: Any.Type, parameters: [String: Any?]) -> PigeonError {
    return PigeonError(
      code: "ConstructorReturnedNullError",
      message: "Failed to instantiate `\(String(describing: type))` with parameters: \(parameters)",
      details: nil)
  }
}

/// Implementation of `WebKitLibraryPigeonProxyApiDelegate` that provides each ProxyApi delegate implementation.
open class ProxyAPIDelegate: CrossFileDarwinApisPigeonProxyApiDelegate {
  func pigeonApiFileHandle(_ registrar: CrossFileDarwinApisPigeonProxyApiRegistrar)
    -> PigeonApiFileHandle
  {
    return PigeonApiFileHandle(pigeonRegistrar: registrar, delegate: FileHandleProxyAPIDelegate())
  }
}
